Finish KIC plugin pages (#1231)
Kong/developer.konghq.com: app/_how-tos/kic-plugin-mtls.md

---
title: mTLS

> description: "Configure the {{ site.kic_product_name }} to verify client certificates using CA certificates and mtls-auth plugin for HTTPS requests."

Author: `mtls-auth` is the name of the plugin

tools:
  - kic

works_on:
  - on-prem
  - konnect

entities: []

tldr:
  q: How do I enforce mTLS from a client to {{ site.base_gateway }} using {{ site.kic_product_name }}?
  a: Create a Secret containing a CA Certificate and pass the ID of the certificate to an mTLS plugin configuration

prereqs:
  enterprise: true
  kubernetes:
    gateway_api: true
  entities:
    services:
      - echo-service
    routes:
      - echo

cleanup:
  inline:
    - title: Uninstall KIC from your cluster
      include_content: cleanup/products/kic
      icon_url: /assets/icons/kubernetes.svg
---

## About mTLS

Mutual TLS (mTLS) is a way to secure connectivity using certificates. {{ site.base_gateway }} can look for a certificate in incoming requests and reject the request if the public key presented does not match the private key stored in {{ site.base_gateway }}.

## Generate a CA certificate

To use the `mtls-auth` plugin you need a CA certificate. If you don't have one, generate a new certificate using `openssl`:

```bash
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -sha256 -days 365 -nodes \
-subj "/C=US/ST=California/L=San Francisco/O=Kong/OU=Org/CN=www.example.com"
```

> ## Add the certificate to {{ site.base_gateway }}

Author: ```suggestion
## Add the Certificate to {{ site.base_gateway }}
```

CA certificates in Kong are provisioned by creating `Secret` or `ConfigMap` resource in Kubernetes.

Resources holding CA certificates must have the following properties:
- the `konghq.com/ca-cert: "true"` label applied.
- a `cert` or `ca.crt` data property which contains a valid CA certificate in PEM format.
- a `kubernetes.io/ingress.class` annotation whose value matches the value of the controller's `--ingress-class` argument. By default, that value is `kong`.
- an `id` data property which contains a random UUID.

Each CA certificate that you create needs a unique ID. Any random UUID should suffice here and it doesn't have a security implication. You can use [uuidgen](https://linux.die.net/man/1/uuidgen) (Linux, OS X) or [New-Guid](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/new-guid) (Windows) to generate an ID.

```bash
CERT_ID=$(uuidgen | tr "[:upper:]" "[:lower:]")
kubectl create secret -n kong generic my-ca-cert --from-literal=id=$CERT_ID --from-file=cert=./cert.pem
kubectl label secret -n kong my-ca-cert 'konghq.com/ca-cert=true'
kubectl annotate secret -n kong my-ca-cert 'kubernetes.io/ingress.class=kong'
```

> ## Configure the mtls-auth plugin
> 
> The [mtls-auth plugin](/plugins/mtls-auth/) requires a CA Certificate ID that will be used to validate the certificate in the incoming request. In this example we disable revocation checks, but you should enable checks in a production setting.

Reviewer: ```suggestion
## 3. Configure the mTLS Auth plugin

The [mTLS Auth plugin](/plugins/mtls-auth/) requires a CA Certificate ID that will be used to validate the Certificate in the incoming request. In this example, we disable revocation checks, but you should enable checks in a production setting.
```

> The [mtls-auth plugin](/plugins/mtls-auth/) requires a CA Certificate ID that will be used to validate the certificate in the incoming request. In this example we disable revocation checks, but you should enable checks in a production setting.

Author: ```suggestion
The [mtls-auth plugin](/plugins/mtls-auth/) requires a CA Certificate ID that will be used to validate the Certificate in the incoming request. In this example we disable revocation checks, but you should enable checks in a production setting.
```

{% entity_example %}
type: plugin
data:
  name: mtls-auth
  config:
    ca_certificates:
    - $CERT_ID
    skip_consumer_lookup: true
    revocation_check_mode: SKIP

  service: echo

> {% endentity_example %}

Reviewer: This didn't work for me, I don't think my enterprise license set correctly. I tried running the cleanup and starting over, but that didn't fix it.

```
Error from server: error when creating "STDIN": admission webhook "validations.kong.konghq.com" denied the request: plugin failed schema validation: schema violation (name: 'mtls-auth' is an enterprise only plugin)
```

> ## Validate your configuration

Reviewer: ```suggestion
## 4. Validate your configuration
```

> At this point {{ site.base_gateway }} will reject requests that do not contain a client certificate.

Reviewer: ```suggestion
At this point, {{ site.base_gateway }} will reject requests that do not contain a client certificate.
```

> 1. Send a request to check Kong prompts you for client certificate.

Reviewer: ```suggestion
1. Send a request to check {{site.base_gateway}} prompts for a client certificate:
```

{% validation request-check %}
url: /echo
status_code: 401
message: No required TLS certificate was sent
on_prem_url: $PROXY_IP
konnect_url: $PROXY_IP
indent: 4
{% endvalidation %}

    As you can see, {{ site.base_gateway }} is restricting the request because it doesn't have the necessary authentication information.

   Two things to note here:

>    - `-k` is used because {{ site.base_gateway }} is set up to serve a self-signed certificate by default. For full mutual authentication in production use cases, you must configure Kong to serve a certificate that is signed by a trusted CA.
>    - For some deployments `$PROXY_IP` might contain a port that points to `http` port of Kong. In others, it might happen that it contains a DNS name instead of an IP address. If needed, update the command to send an `https` request to the `https` port of Kong or the load balancer in front of it.

Reviewer: ```suggestion
   - `-k` is used because {{ site.base_gateway }} is set up to serve a self-signed certificate by default. For full mutual authentication in production use cases, you must configure {{ site.base_gateway }} to serve a Certificate that is signed by a trusted CA.
   - For some deployments `$PROXY_IP` might contain a port that points to `http` port of {{ site.base_gateway }}. In others, it might contain a DNS name instead of an IP address. If needed, update the command to send an `https` request to the `https` port of {{ site.base_gateway }} or the load balancer in front of it.
```

> 1. Use the key and certificate to authenticate against Kong and use the service:

Reviewer: ```suggestion
1. Use the key and Certificate to authenticate against {{ site.base_gateway }} and use the Service:
```
